### PR TITLE
Prevent unexpected file overwrites for collections, drafts and pages

### LINKED ISSUE #489: Creating a new post will overwrite an existing post when the file already exists
### Description:

Creating a new post will overwrite an existing post sometimes.

This happens when the filename where the post is to be saved is the same as an existing filename.

### Tell us a bit about yourself:
* Version of JekyllAdmin I'm using <HINT: use `bundle show` to check>: jekyll-admin (0.8.1)
* Version of Jekyll I'm using <HINT: use `bundle show` to check>: jekyll (3.8.5)
* Version of NodeJS I'm using <HINT: use `node -v` to check>: none
* Operating System <e.g. OS X, Windows>: linux
* Browser <e.g, Safari, Chrome>: Chrome

### Steps to reproduce:

1. Create a new (first) post with the title as "test" and no path.
2. Add some content to the first post.
3. Create the first post.
4. Then, create another new (second) post also with the title "test" and no path. This should be done close enough in time to the first one, i.e: the same day usually.
5. Add some different content to the second post.
6. Create the second post.

### I expected the following:

The second post should not not overwrite the first post.

### But got the following, instead:

The first post will be replaced by the second post, and the first post is lost.

### Other details:

This isn't quite so bad with posts as the filename usually contains a date, so it's less likely to accidentally happen. 

With collections though which the title can be the only element that ends up in the filename, it's easier to overwrite an existing item in a collection. (repeat steps above with a collection other than posts that doesn't have a date in the file name / path.)

Overwriting the file makes sense for editing existing posts, but I don't think it makes sense when creating new posts. (Same for collections.)

### PR DESCRIPTION
**Pull request by alzeih**
For collections, drafts and pages, return a 404 if a new or renamed item will overwrite an existing item.

Fixes #489 .

Let me know if this looks like the right approach.

Two existing tests were altered to match what I was seeing during manual testing. New tests not written yet, I'll wait until I get some feedback on the changes.

### CODE REVIEW

File: lib/jekyll-admin/path_helper.rb [hunk 1]
```diff
@@ -51,6 +51,13 @@ def renamed?
       ensure_leading_slash(request_payload["path"]) != relative_path
     end
 
+    # Is this request creating a new file?
+    def new?
+      return false if request_payload["path"]
+
+      ensure_leading_slash(request_payload["path"]) != relative_path
+    end
+
```
Comment (ashmaroli) [inline lib/jekyll-admin/path_helper.rb]: I'm a bit *unsure* about the duplicated logic in `:new?` and `:renamed?`..

Both methods ultimately test `ensure_leading_slash(request_payload["path"]) != relative_path` explicitly. But in the server namespaces, you have `ensure_not_written_file if new? || renamed?`.
So in `:new?` we test if `request_payload["path"]` is a truthy, and if so immediately call `:renamed?` method where `request_payload["path"]` is again tested for a truthy and then proceed if so..

Needs a bit of a DRYing up and refactoring, IMO.

**Comment by ashmaroli** [comment]
@alzeih If you wish to have this shipped in the upcoming release, please find time to make the requested changes a.s.a.p.
Thank you.

File: lib/jekyll-admin/path_helper.rb (hunk 1 above)
Comment (alzeih) [reply lib/jekyll-admin/path_helper.rb]: I have replaced

```
def new?
  return false if request_payload["path"]

  ensure_leading_slash(request_payload["path"]) != relative_path
end
```
```
def new?
  !request_payload["path"]
end
```

As `new?` doesn't need to compare the value, it only checks that it exists. So this becomes DRY for free.

**Comment by alzeih** [comment]
> * Needs a bit of a refactoring..

Done.

> * Are resources at the `/data/` namespace immune to the issue faced by other types?

No, they are not, but the front end requests for data are inconsistent with those of collections, drafts, and pages, as it doesn't send the path. I can see how to fix it, but it's a separate PR.

**Comment by ashmaroli** [comment]
@alzeih I'm on board with the current approach. However, I think the intent would be better expressed if the `public` methods were named along the lines of `:ensure_not_existing_file` instead.

If you can add more tests, the better.

**Pull request merged**